src/lib/Shop.js

```diff
@@ -339,74 +360,119 @@ class AutomationShop
-                    }.bind(this);
-            }
-            else if ((item.name == "Lucky_egg") || (item.name == "Token_collector"))
-            {
-                itemData.isUnlocked = function() { return TownList["Pewter City"].isUnlocked(); };
-                itemData.isPurchasable = function()
+                        sellableItems[item.name] = {item: item, towns: [townName], requirements: []};
+                    }
+                    else
                     {
-                        return this.__internal__isPokeMarkUnlocked()
-                            || (TownList["Pewter City"].isUnlocked() && Automation.Utils.Route.canMoveToRegion(0))
-                            || TownList["Violet City"].isUnlocked();
-                    }.bind(this);
-            }
-            else if ((item.name == "Dowsing_machine") || (item.name == "Lucky_incense"))
-            {
-                itemData.isUnlocked = function() { return TownList["Lavender Town"].isUnlocked(); };
-                itemData.isPurchasable = function()
+                        sellableItems[item.name].towns.push(townName);
+                    }
+
+                    // Add the requirements
+                    if (shop.requirements !== undefined && shop.requirements.length > 0)
                     {
-                        return this.__internal__isPokeMarkUnlocked()
-                            || (TownList["Lavender Town"].isUnlocked() && Automation.Utils.Route.canMoveToRegion(0))
-                            || TownList["Olivine City"].isUnlocked();
-                    }.bind(this);
-            }
-            else if (item.name == "SmallRestore")
-            {
-                itemData.isUnlocked = function() { return TownList["Cinnabar Island"].isUnlocked(); };
-                itemData.isPurchasable = itemData.isUnlocked;
+                        sellableItems[item.name].requirements.push(shop.requirements);
+                    }
+                    else
+                    {
+                        sellableItems[item.name].requirements.push([]);
+                    }
+                }
+
             }
-            else if (item.name == "MediumRestore")
+        }
+
+        // Iterate with of instead of Object.entries to keep the order
+        for (const itemData of Object.values(sellableItems))
+        {
+            let shopItem = { item: itemData.item }
+            shopItem.isUnlocked = function() {
+                return itemData.towns.some(townName => TownList[townName].isUnlocked())
+                    && (itemData.requirements.length == 0
+                        || itemData.requirements.some(requirementList => requirementList.every(requirement => requirement.isCompleted())));
+            };
+
+            // This may restrict some items that are actually purchasable, but only until the player unlocks the port again
+            shopItem.isPurchasable = () => shopItem.isUnlocked && Automation.Utils.Route.canMoveToRegion(0);
```

Comment:
It would be more accurate, and usable to have a method that ensures that: 
  - At least one unlocked town sells the item
  - The player can move to this town (using the `canMoveToRegion` on the town's one)

This would not be an acceptable regression

Reply:
> This would not be an acceptable regression
It barely affects any amount of time, but fair.

I will add the proper check (so for every town, it has to be unlocked, and should be able to move to that region)

It would still lock items that are always available through the PokéMart outside the map, so I will check there as well

Reply:
> I will add the proper check (so for every town, it has to be unlocked, and should be able to move to that region)

Any, not every ;) Only one available and accessible town is enough.

> It would still lock items that are always available through the PokéMart outside the map, so I will check there as well

That would be appreciated

Reply:
> Any, not every

True, typed too fast here. That is the check that is done in isUnlocked as well.